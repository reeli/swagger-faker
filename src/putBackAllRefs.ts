import { pickRefKey } from "./utils";
import { mapValues, isArray, map } from "lodash";
import { ISchema, IReference } from "./__types__/OpenAPI";

export const putBackAllRefs = (schemas: { [key: string]: ISchema | IReference }) => {
  const store: { [key: string]: any } = {};

  mapValues(schemas, (s, k) => {
    if (store[k]) {
      return store[k];
    }

    const v = putBackRefs(store, schemas, s);
    store[k] = v;
    return v;
  });

  return store;
};

export const putBackRefs = (
  store: any,
  schemas: { [key: string]: ISchema | IReference },
  resp?: ISchema | IReference,
) => {
  if (!resp) {
    return;
  }

  function pickRefData(s: ISchema | IReference) {
    const refKey = pickRefKey(s.$ref);
    if (store[refKey]) {
      return store[refKey];
    }

    const resolvedRef = putBack(schemas[refKey]);
    store[refKey] = resolvedRef;
    return resolvedRef;
  }

  const putBack = (item: ISchema | IReference = {}): any => {
    // handle ref
    if (item.$ref) {
      return pickRefData(item);
    }

    const schema = item as ISchema;
    // handle object
    if (schema.type === "object" || schema.properties) {
      return {
        ...schema,
        properties: mapValues(schema.properties, (v3) => putBack(v3)),
      };
    }

    // handle array
    if (schema.items) {
      if (isArray(schema.items)) {
        return {
          ...schema,
          items: map(schema.items, (v) => putBack(v)),
        };
      }
      return {
        ...schema,
        items: putBack(schema.items),
      };
    }

    // handle oneOf
    if ((schema as any).oneOf) {
      return map((schema as any).oneOf, (v) => putBack(v));
    }

    // handle allOf
    if (schema.allOf) {
      return map((schema as any).allOf, (v) => putBack(v)).reduce((res, item1) => {
        return {
          ...res,
          properties: {
            ...res.properties,
            ...item1.properties,
          },
        };
      }, {});
    }

    return schema;
  };

  return putBack(resp);
};
