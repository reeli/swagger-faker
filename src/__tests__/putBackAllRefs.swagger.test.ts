import { putBackAllRefs } from "../putBackAllRefs";
import swagger from "./mock-data/swagger.json";

describe("should pick all $ref value and put them back", () => {
  it("should pick all $ref value and put them back from swagger", () => {
    expect(putBackAllRefs((swagger as any).definitions)).toMatchSnapshot();
  });
});
