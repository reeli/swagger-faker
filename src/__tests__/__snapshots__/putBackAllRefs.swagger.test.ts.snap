// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should pick all $ref value and put them back should pick all $ref value and put them back from swagger 1`] = `
Object {
  "ApiResponse": Object {
    "properties": Object {
      "code": Object {
        "format": "int32",
        "type": "integer",
      },
      "message": Object {
        "type": "string",
      },
      "type": Object {
        "type": "string",
      },
    },
    "type": "object",
  },
  "Category": Object {
    "properties": Object {
      "id": Object {
        "example": 222,
        "format": "int64",
        "type": "integer",
      },
      "name": Object {
        "example": "Jay",
        "type": "string",
      },
    },
    "type": "object",
    "xml": Object {
      "name": "Category",
    },
  },
  "Order": Object {
    "properties": Object {
      "complete": Object {
        "default": false,
        "type": "boolean",
      },
      "id": Object {
        "format": "int64",
        "type": "integer",
      },
      "petId": Object {
        "format": "int64",
        "type": "integer",
      },
      "quantity": Object {
        "format": "int32",
        "type": "integer",
      },
      "shipDate": Object {
        "format": "date-time",
        "type": "string",
      },
      "status": Object {
        "description": "Order Status",
        "enum": Array [
          "placed",
          "approved",
          "delivered",
        ],
        "type": "string",
      },
    },
    "type": "object",
    "xml": Object {
      "name": "Order",
    },
  },
  "Pet": Object {
    "properties": Object {
      "category": Object {
        "properties": Object {
          "id": Object {
            "example": 222,
            "format": "int64",
            "type": "integer",
          },
          "name": Object {
            "example": "Jay",
            "type": "string",
          },
        },
        "type": "object",
        "xml": Object {
          "name": "Category",
        },
      },
      "id": Object {
        "example": 111,
        "format": "int64",
        "type": "integer",
      },
      "name": Object {
        "example": "doggie",
        "type": "string",
      },
      "photoUrls": Object {
        "example": Array [
          "a.png",
          "b.png",
        ],
        "items": Object {
          "type": "string",
        },
        "type": "array",
        "xml": Object {
          "name": "photoUrl",
          "wrapped": true,
        },
      },
      "status": Object {
        "description": "pet status in the store",
        "enum": Array [
          "available",
        ],
        "type": "string",
      },
      "tags": Object {
        "example": Array [
          Object {
            "id": 111,
            "name": "Lucy",
          },
        ],
        "items": Object {
          "properties": Object {
            "id": Object {
              "format": "int64",
              "type": "integer",
            },
            "name": Object {
              "type": "string",
            },
          },
          "type": "object",
          "xml": Object {
            "name": "Tag",
          },
        },
        "type": "array",
        "xml": Object {
          "name": "tag",
          "wrapped": true,
        },
      },
    },
    "required": Array [
      "name",
      "photoUrls",
    ],
    "type": "object",
    "xml": Object {
      "name": "Pet",
    },
  },
  "Tag": Object {
    "properties": Object {
      "id": Object {
        "format": "int64",
        "type": "integer",
      },
      "name": Object {
        "type": "string",
      },
    },
    "type": "object",
    "xml": Object {
      "name": "Tag",
    },
  },
  "User": Object {
    "properties": Object {
      "email": Object {
        "type": "string",
      },
      "firstName": Object {
        "type": "string",
      },
      "id": Object {
        "format": "int64",
        "type": "integer",
      },
      "lastName": Object {
        "type": "string",
      },
      "password": Object {
        "type": "string",
      },
      "phone": Object {
        "type": "string",
      },
      "userStatus": Object {
        "description": "User Status",
        "format": "int32",
        "type": "integer",
      },
      "username": Object {
        "type": "string",
      },
    },
    "type": "object",
    "xml": Object {
      "name": "User",
    },
  },
}
`;
